### ISSUE TITLE
[Bug] tensor.gather on local TileType input can overwrite source row during lowering

### ISSUE THREAD

**Issue by zhaozhaozz** [hw-native-sys/pypto#1622]
### Component

IR / `ConvertTensorToTileOps` / `tensor.gather` lowering for `TileType` inputs

### Description

`tensor.gather` can mutate or alias its local TileType input when the input is a local tensor lowered to a tile.

The repro below creates a local `[16, 256]` FP32 tensor `local_scores`, reads it with a scalar loop, then calls:

```python
gathered_tile = pl.gather(local_scores, dim=-1, index=idx)
```

The gather output itself is correct. However, reading the same `local_scores` again with the same scalar loop after `pl.gather` returns wrong values: row 0 behaves as if it has been overwritten by row 15.

This violates the expected read-only semantics of `gather`: neither the source input nor the index input should be modified by the operation.

This was observed on PyPTO main after #1609, where `tensor.gather` now accepts `TileType` input/index and no longer fails at compile time.

### Minimal Reproducer

Save this as `repro_tensor_gather_local_alias.py` and run it with the normal PyPTO runtime environment configured.

```python
import argparse

import pypto.language as pl
import torch
from pypto.runtime.runner import RunConfig


ROWS = 16
COLS = 256
TOPK = 8


@pl.jit
def repro_kernel(
    scores: pl.Tensor[[ROWS, COLS], pl.FP32],
    idx: pl.Tensor[[ROWS, TOPK], pl.INT32],
    before: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
    gathered: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
    after: pl.Out[pl.Tensor[[ROWS, TOPK], pl.FP32]],
):
    for _ in pl.spmd(1, name_hint="gather_alias_repro"):
        local_scores = pl.create_tensor([ROWS, COLS], dtype=pl.FP32)
        local_scores[:, :] = scores[:, :]

        local_idx = pl.create_tensor([ROWS, TOPK], dtype=pl.INT32)
        local_idx[:, :] = idx[:, :]

        before_tile = pl.full([ROWS, TOPK], dtype=pl.FP32, value=0.0)
        for r in pl.range(ROWS):
            for k in pl.range(TOPK):
                col_i32 = pl.read(local_idx, [r, k])
                col = pl.cast(col_i32, pl.INDEX)
                val = pl.read(local_scores, [r, col])
                pl.write(before_tile, [r, k], val)
        before[:, :] = before_tile

        gathered_tile = pl.gather(local_scores, dim=-1, index=idx)
        gathered[:, :] = gathered_tile

        after_tile = pl.full([ROWS, TOPK], dtype=pl.FP32, value=0.0)
        for r in pl.range(ROWS):
            for k in pl.range(TOPK):
                col_i32 = pl.read(local_idx, [r, k])
                col = pl.cast(col_i32, pl.INDEX)
                val = pl.read(local_scores, [r, col])
                pl.write(after_tile, [r, k], val)
        after[:, :] = after_tile


def make_scores() -> torch.Tensor:
    row = torch.arange(ROWS, dtype=torch.float32).reshape(ROWS, 1) * 1000.0
    col = torch.arange(COLS, dtype=torch.float32).reshape(1, COLS)
    return row + col


def make_idx() -> torch.Tensor:
    base = torch.tensor([7, 9, 3, 128, 255, 0, 64, 11], dtype=torch.int32)
    return torch.stack([(base + r * 13) % COLS for r in range(ROWS)])


def first_mismatch(actual: torch.Tensor, expected: torch.Tensor) -> str:
    neq = actual != expected
    if not bool(neq.any()):
        return ""
    flat_idx = int(torch.nonzero(neq.reshape(-1), as_tuple=False)[0].item())
    idx = torch.unravel_index(torch.tensor(flat_idx), actual.shape)
    idx_tuple = tuple(int(x) for x in idx)
    return f"{idx_tuple}: actual={actual[idx_tuple].item()} expected={expected[idx_tuple].item()}"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("-p", "--platform", default="a2a3sim", choices=["a2a3", "a2a3sim", "a5", "a5sim"])
    parser.add_argument("-d", "--device", type=int, default=0)
    args = parser.parse_args()

    scores = make_scores()
    idx = make_idx()
    expected = torch.gather(scores, dim=-1, index=idx.long())
    before = torch.zeros_like(expected)
    gathered = torch.zeros_like(expected)
    after = torch.zeros_like(expected)

    repro_kernel(
        scores,
        idx,
        before,
        gathered,
        after,
        config=RunConfig(platform=args.platform, device_id=args.device, dump_passes=True),
    )

    for name, actual in {"before": before, "gathered": gathered, "after": after}.items():
        ok = torch.equal(actual, expected)
        print(f"{name}: {'PASS' if ok else 'FAIL'}")
        if not ok:
            print(f"  first mismatch: {first_mismatch(actual, expected)}")


if __name__ == "__main__":
    main()
```

Run:

```bash
python repro_tensor_gather_local_alias.py -p a2a3sim
```

### Expected Behavior

All three outputs should match `torch.gather(scores, dim=-1, index=idx)`:

```text
before: PASS
gathered: PASS
after: PASS
```

`before` and `after` use the same scalar-read loop:

```python
val = pl.read(local_scores, [r, col])
pl.write(out_tile, [r, k], val)
```

The only operation between them is `pl.gather(local_scores, dim=-1, index=idx)`, which should not mutate `local_scores`.

### Actual Behavior

Observed result:

```text
before: PASS
gathered: PASS
after: FAIL
  first mismatch: (0, 0): actual=15007.0 expected=7.0
```

The input is constructed as:

```python
scores[row, col] = row * 1000 + col
```

Therefore `15007.0` means the `after` scalar-read path read row 15, col 7 where row 0, col 7 was expected. The scalar-read loop itself is not the cause, because the same loop passes before `pl.gather`.

### Preliminary Analysis

The generated PTO shows that the TileType-input `tensor.gather` path materializes each source row into a scratch tile before calling `pto.tgather`. The scratch tile for the sliced `[1, 256]` row aliases the same base address as the full `local_scores` tile:

```mlir
%local_scores__tile = pto.alloc_tile addr = %c96_i64 ... rows=16, cols=256
...
%0 = pto.alloc_tile addr = %c96_i64 ... rows=16, cols=256
...
%gather_inp_row = pto.alloc_tile addr = %c96_i64 ... rows=1, cols=256
%slice_view = pto.subview %0[%gather_lv, %c0_index] sizes [1, 256] ...
pto.tmov ins(%slice_view) outs(%gather_inp_row)
pto.tgather ins(%gather_inp_row, %gather_idx_row, %gather_row_tmp) outs(%gather_row)
```

Since `%gather_inp_row` starts at the same address as `local_scores`, every row materialization writes the current source row into `local_scores` row 0. The gather result is still correct in the same iteration because `pto.tgather` immediately consumes that scratch row. After the loop finishes, row 0 has been overwritten by the last materialized row, row 15. This explains the `after` mismatch:

```text
after[0, 0] reads local_scores[0, 7]
actual = 15007.0  # row 15, col 7
expected = 7.0    # row 0, col 7
```

This suggests that the TileType-input `tensor.gather` lowering needs to ensure its row materialization scratch does not alias the source input tile, or otherwise avoid materializing by writing into the source tile's address range.

### Environment

Observed with:

```text
PyPTO: b5341e805fae2c4ee0b3ec1615b59755fa1015f9
ptoas: 0.42
platform: a2a3sim
```

### Related

- #1609 fixed the compile-time `TileType` acceptance path for `tensor.gather`.
- This issue tracks a separate runtime correctness problem after the code can compile and execute.

**Comment by zhaozhaozz** [comment]
### Additional confirmed problem path

This comment only records the confirmed path that mutates the source tile in the reproducer. It does not propose a final fix.

#### 1. `tensor.gather` emits a `tile.slice` with an explicit `valid_shape` for TileType input rows

For `tensor.gather(input, dim=-1, index=idx)` when `input` is already a `TileType`, the conversion path in `src/ir/transforms/op_conversion_registry.cpp` emits a per-row `tile.slice` for the source row. In the current code, that helper builds the slice with four arguments:

```cpp
tile.slice(src, shape, offsets, shape)
```

The fourth argument is an explicit `valid_shape` and is identical to `shape` in this path.

#### 2. Four-argument `tile.slice` takes the materializing PTO codegen path

In `src/backend/common/pto_ops_common.cpp`, `tile.slice` checks whether an explicit `valid_shape` operand exists:

```cpp
bool has_explicit_valid_shape = valid_tuple != nullptr && !valid_tuple->elements_.empty();
```

When `has_explicit_valid_shape` is true and the slice is not handled by the full-window fast path, codegen emits:

```mlir
%slice_view = pto.subview ...
pto.tmov ins(%slice_view ...) outs(%gather_inp_row ...)
pto.set_validshape %gather_inp_row, ...
```

So the four-argument `tile.slice` path does not leave the result as a pure `pto.subview`; it materializes the subview into the slice result buffer.

#### 3. `tile.slice` result MemRef is still a view of the source MemRef

`tile.slice` is registered as an output-memory-inherits-input operation, so `InitMemRef` routes it through `ShareMemRefFrom` in `src/ir/transforms/init_memref.cpp`.

That path creates the slice result MemRef from the source MemRef base, plus the slice byte offset and the output view size. Therefore the slice result still belongs to the same allocation base as the source tile.

`src/ir/transforms/allocate_memory_addr_pass.cpp` documents the dynamic-offset case: if a view offset is dynamic, the allocated MemRef address falls back to the base address, and pure `tile.slice` codegen is expected to derive the dynamic offset through `pto.subview` operands.

#### 4. The failing generated PTO writes the materialized row into the source allocation base

In the PTO emitted by the reproducer, the source tile allocation and the materialized row allocation share the same address:

```mlir
%local_scores__tile = pto.alloc_tile addr = %c96_i64 ... rows=16, cols=256
...
%gather_inp_row = pto.alloc_tile addr = %c96_i64 ... rows=1, cols=256
%slice_view = pto.subview %0[%gather_lv, %c0_index] sizes [1, 256] ...
pto.tmov ins(%slice_view ...) outs(%gather_inp_row ...)
pto.tgather ins(%gather_inp_row, %gather_idx_row, %gather_row_tmp ...) outs(%gather_row ...)
```

Because `%gather_inp_row` starts at the same address as the source tile allocation, the `pto.tmov` materialization writes the current source row into row 0 of the source allocation. After the per-row loop completes, row 0 contains the last materialized row.

This directly matches the minimal reproducer output:

```text
before: PASS
gathered: PASS
after: FAIL
  first mismatch: (0, 0): actual=15007.0 expected=7.0
```

The value `15007.0` corresponds to row 15, col 7 in the reproducer input (`scores[row, col] = row * 1000 + col`). The scalar read after `pl.gather` therefore observes that row 0 has been overwritten by row 15. The same scalar read before `pl.gather` passes, and the gather result itself passes, so the mismatch is introduced by the materializing row-slice path between those two reads.
